File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        # min len starting at the end would be at least one
        # so one is the least LIS hence init cache with 1
        res = [1] * len(nums)

        """
        - - - - -
        """


        # start at the end of the seq and go back towards start
        for i in range(len(nums)-1, -1, -1):
            # at each index we will compare the LIS till the end of list
            for j in range(i+1, len(nums)):
                # compare if the two nums are in increasing order
                if nums[i] < nums[j]:
                    # len of LIS at the current index
                    res[i] = max(res[i], 1 + res[j])

        return max(res)
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {

        int aryLen = nums.length;
        int[] res = new int[aryLen];

        res[0] = 1;
        for(int i=1; i<aryLen; i++) {
            res[i] = res[i-1] * nums[i-1];
        }

        int postFix = 1;
        for(int i = aryLen-1; i >=0; i--) {
            res[i] *= postFix;
            postFix *= nums[i];
        }

        return res;
    }
}  
